src/server/scripts/Kalimdor/Firelands/boss_ragnaros_firelands.cpp

```diff
@@ -0,0 +1,5304 @@
+            {
+                switch (summon->GetEntry())
+                {
+                    case NPC_SON_OF_FLAME:
+                        summons.Despawn(summon);
+                        if (events.IsInPhase(PHASE_INTERMISSION) && !summons.HasEntry(NPC_SON_OF_FLAME))
+                            PreparePhase(_previousPhase == PHASE_ONE ? PHASE_TWO : PHASE_THREE);
+                        break;
+                    case NPC_LAVA_SCION:
+                        if (instance)
+                            instance->SendEncounterUnit(ENCOUNTER_FRAME_DISENGAGE, summon);
+                        break;
+                    default:
+                        break;
+                }
+            }
+
+            void SummonedCreatureDespawn(Creature* summon) override
+            {
+                switch (summon->GetEntry())
+                {
+                    case NPC_LAVA_SCION:
+                        if (instance)
+                            instance->SendEncounterUnit(ENCOUNTER_FRAME_DISENGAGE, summon);
+                        break;
+                    default:
+                        break;
+                }
+
+                if (summon->IsAIEnabled)
+                    summon->AI()->EnterEvadeMode();
+
+                BossAI::SummonedCreatureDespawn(summon);
+            }
+
+            void DoAction(int32 action) override
+            {
+                switch (action)
+                {
+                    case ACTION_RAGNAROS_HEROIC_SUBMERGE:
+                        Talk(SAY_RAGNAROS_HEROIC_SUBMERGE);
+                        me->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NOT_SELECTABLE);
+                        me->CastSpell(me, SPELL_LEGS_SUBMERGE, true);
+                        me->RemoveAurasDueToSpell(SPELL_BASE_VISUAL);
+                        me->AttackStop();
+                        me->SetReactState(REACT_PASSIVE);
+                        events.ScheduleEvent(EVENT_HEROIC_CENARIUS_1, 3.2 * IN_MILLISECONDS);
+                        events.ScheduleEvent(EVENT_PREPARE_PHASE_FOUR, 5.2 * IN_MILLISECONDS);
+                        break;
+                    case ACTION_RAGNAROS_DREADFLAME:
+                        _dreadflameCountdown -= std::min(_dreadflameCountdown, uint32(DREADFLAME_TICK));
+                        if (_dreadflameCountdown != 0 && _dreadflameCnt > 0)
+                            break;
+
+                        _dreadflameCountdown = DREADFLAME_BASE_CD - (_dreadflameCnt * DREADFLAME_TICK);
+                        ++_dreadflameCnt;
+                        if (Creature* dreadflame = ObjectAccessor::GetCreature(*me, instance->GetGuidData(NPC_DREADFLAME)))
+                        {
+                            DreadflameAI* ai = CAST_AI(DreadflameAI, dreadflame->GetAI());
+                            if (ai)
+                            {
+                                std::list<Position> positions = ai->GetRandomVacantDreadflamePositions(me->GetMap()->Is25ManRaid() ? 6 : 2);
+  
```

Comment:
please dont try to write floating_point * IN_MILLISECONDS, I guess you did it because it looks nicer but it also causes a compiler warning about conversion from double to uint32 (which you would need a cast to silence and at this point it just stops looking nice)